rspace/src/main/scala/coop/rchain/rspace/merger/instances/DiffStateMerger.scala

```diff
@@ -0,0 +1,317 @@
+                        endStateReader.getRichContinuations
+                      )
+                  // populate main state content store with continuations on this key
+                  _ <- mainConsumes.get(consumeHash)
+                  _ <- cDiffRef.update(s => {
+                        val curVal = s.getOrElse(consumeHash, ChannelChange())
+                        val newVal = curVal.copy(
+                          added = curVal.added ++ r.added,
+                          removed = curVal.removed ++ r.removed
+                        )
+                        s.updated(consumeHash, newVal)
+                      })
+                } yield ())
+          )
+          // compute cumulative joins on channel across all states that are merged
+          // during accumulation joins can be duplicated
+          joinsComputes = consumeMappings
+            .flatMap {
+              case ConsumeMapping(_, channelsInvolved) => channelsInvolved
+            }
+            .map {
+              channelHash =>
+                fs2.Stream
+                  .eval(
+                    for {
+                      joinHash    <- joinLeafForChannel.get(channelHash)
+                      joinAtMain  <- mainJoins.get(joinHash)
+                      joinAtMerge <- endStateReader.getRichJoins(joinHash)
+                      _ <- jAccRef.update(s => {
+                            val curVal = s.getOrElse(joinHash, joinAtMain)
+                            val newVal = (curVal ++ joinAtMerge).distinct
+                            s.updated(joinHash, newVal)
+                          })
+                    } yield ()
+                  )
+            }
+
+          // compute data, continuation changes and cumulative joins
+          _ <- Stopwatch.time(Log[F].info(_))(
+                s"compute changes: ${produceChangeComputes.size} produces + ${consumeChangeComputes.size} consumes"
+              )(
+                fs2.Stream
+                  .emits(produceChangeComputes ++ consumeChangeComputes ++ joinsComputes)
+                  .parJoinProcBounded
+                  .compile
+                  .drain
+              )
+        } yield ())
+      }))
+
+      // compute all changes
+      _ <- work.parJoinProcBounded.compile.drain
+
+      // get all data computed
+      initData           <- mainProduces.toMap
+      initCons           <- mainConsumes.toMap
+      initJoins          <- mainJoins.toMap
+      joinForChannelMap  <- joinLeafForChannel.toMap
+      produceChanges     <- pDiffRef.get
+      consumeChanges     <- cDiffRef.get
+      joinsCumulative    <- jAccRef.get
+      channelsPerConsume <- channelsPerConsumesRef.get
+
+      // transform changes into actions over trie
+      produceActions = produceChanges.toList.par.map {
+        case (dataHash, changes) =>
+          val newDatumsAtProduce =
+            (initData(dataHash) ++ changes.added diff changes.removed).distinct
```

Comment:
Why do we need distinct here? Is it because every datum is unique?

Reply:
Another question for these `++` and `diff` in `changes` which is generated by parallel processing. Would the order be changed every time based on different execution? Is that gonna affect the dataHash because of the sequence order?

Reply:
> Another question for these `++` and `diff` in `changes` which is generated by parallel processing. Would the order be changed every time based on different execution? Is that gonna affect the dataHash because of the sequence order?

Oh. I just found the answer that Seq[datum] would be sorted first before encoded by `encodeSortedRich` which can be ensure order which doesnt lead to different hash.

Reply:
`Why do we need distinct here? Is it because every datum is unique?` I feel that you are right, I think there is no need for this distinct. Need to think a more.

Reply:
Yes, you are right! I checked and tests failed if I remove `distinct`. Thats because I do double work - channel can be checked both agains produce and against consume, so I need to have here
https://github.com/rchain/rchain/blob/79d42ba36857599a2fcc75ac8af18ec5b3e71184/rspace/src/main/scala/coop/rchain/rspace/merger/instances/DiffStateMerger.scala#L148-L151

`changedProduceHashes.distinct`.

Reply:
You are right, it should not be `distinct` here. But it should be in another place. We might have duplicate produces/consumes in event log if it is newly introduced persistent and immediately comm-ed. So I will move it to events computation

Reply:
Also if several events touch the same channel, it can lead to duplicate work, so Set should be used when computing channels to check. Please see https://github.com/rchain/rchain/pull/3312/commits/5dac4d19ffbc84e85608dee780419b55009134d8

Reply:
`Is it because every datum is unique?` Answering this - I believe it should be unique, actually, but this is not related to merging logic, so better not assume this.